### PR TITLE
Speedup items processing by caching and skipping repeated checks of item tags

### PR DESCRIPTION
**Pull request by Aivean**
#### Summary

SUMMARY: Performance "Speedup items processing by caching and skipping repeated checks of item tags"

#### Purpose of change

Even after switching to string/int id  (#44694) item tag checks still take lots of CPU time in item-heavy areas, such as labs and hotels.  

Depending on the scenario it can range from 3% to 5+% of game_turn time.

#### Describe the solution

Current list of flags that require processing:
* flag_FAKE_SMOKE
* flag_FAKE_MILL
* is_corpse
* flag_WET
* flag_LITCIG
* flag_WATER_EXTINGUISH
* flag_CABLE_SPOOL
* flag_IS_UPS

It's apparent that most common items don't have these flags. Also, item flags are rarely changed, so these checks could be easily cached.

I've added  `bool item::requires_tags_processing` field that is `true` by default and is set to `true` when any change to item flags is made.

When this field is true, item flags are checked as before. If none of the checks succeed (Important: just checks, not processing!) `requires_tags_processing` is set to `false` and on next iteration flag checks are skipped for this item.

#### Describe alternatives you've considered

Originally I thought of the more complicated solution that involved storing a vector of "flag_processors"  (i.e. pointers to the methods that do processing when flag is present).  

But current solution is simpler and faster.

#### Testing

Made sure that game works. Tried to lit up/extinguish a cigar and checked with debugger that flags processing works as before.

#### Additional context

Profiling results.
Before:
image

After:
image

### CODE REVIEW

**Comment by Aivean** [comment]
> * 8 for Assignment where comparison was intended

ah, but I'm pretty sure I mean assignment there. Well, seems like I'll have to wrap `(mark = true)` in lambda just to silence this.

**Comment by Aivean** [comment]
Some more details on why this cache helps and notes on possible future improvements:

Currently items marked as "active" trigger (relatively) expensive flag checks during the processing.
I found two categories of "active" items that are pretty common:
* food (including water)
* corpses

This PR helps to skip the flag checks for the "food" category. 
This PR is ok as a stopgap measure, but there is room for improvement. 

I think we need to split "active item cache" by "processing type", similarly to `active_item_cache::special_items`. "Processing type" examples include: "food", "corpse", "relic", "ethereal", ... And there might be "catch-all" category "active other". Depending on flags and conditions, item may participate in more than one "processing queues".

Processing logic for each queue will be very simple and targeted, and as the result with this approach we'll avoid all unnecessary "parasitic" checks.

**Comment by kevingranade** [comment]
This cycling priority queue system was basically workable until we hit freeze/thaw mechanics for a broad swath of item types, but I think it's just not going to scale going forward.
What we can do is flip around how we schedule updates.
Now: We guarantee that an item gets processed once per characteristic period, things like explosives must have per-turn resolution and therefore get processed every turn, foodstuff just needs to be processed reasonably often, so once per several minutes.  The problem is millions of items divided by hundreds of turns is still thousands of items per turn, and when cycling rapidly through turns such as when waiting or crafting or sleeping, that delay becomes unreasonable.

My proposed fix for this is to instead have a system where any time an item is processed, the processing code produces an estimate for when it needs to be processed again, and inserts it into a data structure that triggers processing at or before that deadline, for items with a known transition time, this means they will only be processed when loaded and when their time is up, and for items with a varying decay time such as rotting, freezing or thawing, they can asymptotically approach an ideal number of processing events.

In the short term, that processing could be a fixed period identical to what we have now, but then we can pieccewise add better heuristics to different processing logic to reduce the number of processing events per item.

File: src/item.cpp [hunk 1]
```diff
@@ -9934,34 +9938,50 @@ bool item::process_internal( player *carrier, const tripoint &pos,
...
-        return process_UPS( carrier, pos );
+    if( requires_tags_processing ) {
+        // `mark` becomes true if any of the flags that require processing are present
+        bool mark = false;
+        const auto mark_flag = [&mark]() {
+            mark = true;
+            return true;
+        };
+
+        if( has_flag( flag_FAKE_SMOKE ) && mark_flag() && process_fake_smoke( carrier, pos ) ) {
```
Comment (jbytheway) [inline src/item.cpp:9949]: For future reference, I believe using double parens around `mark = true` would also avoid the warning about confusing assignment with comparison (but I haven't actually tested this).

**Pull request merged**